compare/compare.c

```diff
@@ -0,0 +1,627 @@
+    unsigned min  = tm %   60;  tm /=   60;
+    printf("  Time %u:%02u:%02u.%04u - ", (unsigned) tm, min, sec, msec);
+
+    char path[4096];
+    snprintf(path, sizeof(path), "%s/%s", input, file);
+
+    Image16 target;
+    if (!read_png(path, &target)) {
+        printf("PNG reading failed!\n");
+        return false;
+    }
+
+    uint16_t *grad = malloc(2 * target.width * target.height);
+    if (!grad) {
+        free(target.buffer);
+        return out_of_memory();
+    }
+    calc_grad(&target, grad);
+
+    ass_set_frame_size(renderer, scale * target.width, scale * target.height);
+    ASS_Image *img = ass_render_frame(renderer, track, time, NULL);
+
+    const char *out_file = NULL;
+    if (output) {
+        snprintf(path, sizeof(path), "%s/%s", output, file);
+        out_file = path;
+    }
+    double max_err;
+    int res = compare(&target, grad, img, out_file, &max_err, scale);
+    free(target.buffer);
+    free(grad);
+    if (!res)
+        return out_of_memory();
+    bool flag = max_err < 4;
+    printf("%.3f %s\n", max_err, flag ? (max_err < 2 ? "OK" : "BAD") : "FAIL");
+    if (res < 0)
+        printf("Cannot write PNG to file '%s'!\n", path);
+    return flag;
+}
+
+
+typedef struct {
+    char *name;
+    int64_t time;
+} Item;
+
+typedef struct {
+    size_t n_items, max_items;
+    Item *items;
+} ItemList;
+
+static bool init_items(ItemList *list)
+{
+    int n = 256;
+    list->n_items = list->max_items = 0;
+    list->items = malloc(n * sizeof(Item));
+    if (!list->items)
+        return out_of_memory();
+    list->max_items = n;
+    return true;
+}
+
+static bool add_item(ItemList *list)
+{
+    if (list->n_items < list->max_items)
+        return true;
+
+    int n = 2 * list->max_items;
+    Item *next = realloc(list->items, n * sizeof(Item));
+    if (!next)
+        return out_of_memory();
+    list->max_items = n;
+    list->items = next;
+    return true;
+}
+
+static void delete_items(ItemList *list)
+{
+    for (size_t i = 0; i < list->n_items; i++)
+        free(list->items[i].name);
+    free(list->items);
+}
+
+static int item_compare(const void *ptr1, const void *ptr2)
+{
+    const Item *e1 = ptr1, *e2 = ptr2;
+    int cmp = strcmp(e1->name, e2->name);
+    if (cmp)
+        return cmp;
+    if (e1->time > e2->time)
+        return +1;
+    if (e1->time < e2->time)
+        return -1;
+    return 0;
+}
+
+
+static bool add_sub_item(ItemList *list, const char *file, int len)
+{
+    if (!add_item(list))
+        return false;
+
+    Item *item = &list->items[list->n_items];
+    item->name = strndup(file, len);
+    if (!item->name)
+        return out_of_memory();
+    item->time = -1;
+    list->n_items++;
+    return true;
+}
+
+static bool add_img_item(ItemList *list, const char *file, int len)
+{
+    // Parse image name:
+    // <subtitle_name>-<time_in_msec>.png
+
+    int pos = len, first = len;
+    while (true) {
+        if (!pos--)
+            return true;
```

Comment:
Is `true` instead of `false` intentional in lines 425,429,433 ?

Reply:
Yes, `false` means unrecoverable error (allocation failure) that leads to immediate abort. Some stray `png` file in test directory is almost regular case so `true` is intended here.

Reply:
Ok, that works, though a comment documenting the return values might be helpful.
While not strictly necessary, I think it might also be a beneficial to at least print a info about ignored images, to make sure no images are left out on accident.
_(Or alternatively, the test could be more strict and require additional images to be prefixed with a dot, as dotfiles are filtered out anyway in `main`.)_

Reply:
I think that test directory can contain not only automatic test files but some human-readable description too. And that description can have images. Also I think that ability to disable tests by changing extension of `ass`-files only is useful too. I'm reporting all active test images anyway (with total count) so I think that's sufficient as error check at test creation.